File: python/src/reachy_jev/state.py
"""Convert observation numbers to the small vocabulary sent to Jev."""

from __future__ import annotations

import math
import re
from typing import Any

_PERSON = re.compile(r"p[1-9]\Z")


def _finite(value: Any) -> bool:
    return isinstance(value, (int, float)) and not isinstance(value, bool) and math.isfinite(value)


def _optional_bool(value: Any, field: str) -> bool | None:
    if value is None:
        return None
    if not isinstance(value, bool):
        raise ValueError(f"{field} must be boolean")
    return value


def _optional_choice(value: Any, choices: tuple[str, ...], field: str) -> str | None:
    if value is None:
        return None
    if not isinstance(value, str) or value not in choices:
        raise ValueError(f"invalid {field}")
    return value


def _bounded_transcript(value: Any) -> str:
    if not isinstance(value, str):
        raise ValueError("transcript text must be string")
    chars: list[str] = []
    units = 0
    for char in value:
        code_point = ord(char)
        if 0xD800 <= code_point <= 0xDFFF:
            raise ValueError("invalid transcript Unicode")
        width = 2 if code_point > 0xFFFF else 1
        if units + width > 200:
            break
        chars.append(char)
        units += width
    return "".join(chars)


def bearing(degrees: float) -> str:
    if not _finite(degrees):
        raise ValueError("bearing must be finite")
    if degrees < -60:
        return "far left"
    if degrees < -25:
        return "left"
    if degrees < -8:
        return "slightly left"
    if degrees <= 8:
        return "center"
    if degrees <= 25:
        return "slightly right"
    if degrees <= 60:
        return "right"
    return "far right"


def distance(fraction: float) -> str:
    if not _finite(fraction) or not 0 <= fraction <= 1:
        raise ValueError("face height fraction must be in [0,1]")
    return (
        "very near"
        if fraction > 0.35
        else "near"
        if fraction >= 0.20
        else "medium"
        if fraction >= 0.10
        else "far"
    )


def elapsed(seconds: float) -> str:
    if not _finite(seconds) or seconds < 0:
        raise ValueError("elapsed seconds must be non-negative")
    if seconds < 2:
        return "just now"
    if seconds < 10:
        return "a few seconds"
    if seconds < 20:
        return "about 10 seconds"
    if seconds < 45:
        return "about half a minute"
    if seconds < 90:
        return "about a minute"
    return "over a minute"


def sound_level(dbfs: float) -> str:
    if not _finite(dbfs) or dbfs > 0:
        raise ValueError("dBFS must be finite and <= 0")
    return "silent" if dbfs < -55 else "quiet" if dbfs < -35 else "conversational" if dbfs < -15 else "loud"


def build_room_state(observation: dict[str, Any]) -> dict[str, Any]:
    """Accept the same camelCase observation shape as the TypeScript package.

    Missing or non-finite sensor fields are omitted, never inferred. Caller-supplied
    person IDs must be session-local p1..p9 identifiers.
    """
    if not isinstance(observation, dict):
        raise ValueError("invalid room observation shape")
    for field, kind in (("people", list), ("sound", dict), ("robot", dict), ("transcriptRecent", list)):
        value = observation.get(field)
        if value is not None and not isinstance(value, kind):
            raise ValueError("invalid room observation shape")
    people: list[dict[str, Any]] = []
    person_ids: set[str] = set()
    for item in observation.get("people") or []:
        if not isinstance(item, dict):
            raise ValueError("invalid person observation")
        person_id = item.get("id")
        if not isinstance(person_id, str) or not _PERSON.fullmatch(person_id):
            raise ValueError("person IDs must be session-local p1..p9 identifiers")
        if person_id in person_ids:
            raise ValueError("duplicate person ID")
        person_ids.add(person_id)
        never_spoke = _optional_bool(item.get("neverSpoke"), "person.neverSpoke")
        if never_spoke and item.get("secondsSinceLastSpoke") is not None:
            raise ValueError("neverSpoke conflicts with secondsSinceLastSpoke")
        person: dict[str, Any] = {"id": person_id}
        for source, output, convert in (
            ("bearingDeg", "bearing", bearing),
            ("faceHeightFraction", "distance", distance),
            ("secondsSinceLastSpoke", "seconds_since_last_spoke", elapsed),
        ):
            if _finite(item.get(source)):
                person[output] = convert(item[source])
        if _finite(item.get("faceYawDeg")):
            person["facing_robot"] = abs(item["faceYawDeg"]) < 20
        for source, output in (("lookingAtRobot", "looking_at_robot"), ("speaking", "speaking")):
            value = _optional_bool(item.get(source), f"person.{source}")
            if value is not None:
                person[output] = value
        moving = _optional_choice(item.get("moving"), ("still", "shifting", "walking"), "person.moving")
        if moving is not None:
            person["moving"] = moving
        if never_spoke:
            person["seconds_since_last_spoke"] = "never"
        people.append(person)

    state: dict[str, Any] = {"schema": "room_state@1", "people": people}
    sound = observation.get("sound") or {}
    sound_state: dict[str, Any] = {}
    if _finite(sound.get("loudestBearingDeg")):
        sound_state["loudest_bearing"] = bearing(sound["loudestBearingDeg"])
    if _finite(sound.get("levelDbfs")):
        sound_state["level"] = sound_level(sound["levelDbfs"])
    voice_detected = _optional_bool(sound.get("voiceDetected"), "sound.voiceDetected")
    if voice_detected is not None:
        sound_state["voice_detected"] = voice_detected
    if sound_state:
        state["sound"] = sound_state

    counts: dict[str, int] = {}
    recent: list[dict[str, Any]] = []
    for utterance in reversed(observation.get("transcriptRecent") or []):
        if not isinstance(utterance, dict):
            raise ValueError("invalid transcript observation")
        who = utterance.get("who")
        if who != "unknown" and (not isinstance(who, str) or not _PERSON.fullmatch(who)):
            continue
        if counts.get(who, 0) >= 2:
            continue
        counts[who] = counts.get(who, 0) + 1
        entry: dict[str, Any] = {"who": who, "text": _bounded_transcript(utterance["text"])}
        if _finite(utterance.get("endedSecondsAgo")):
            entry["ended"] = elapsed(utterance["endedSecondsAgo"])
        recent.append(entry)
        if len(recent) == 4:
            break
    if recent:
        state["transcript_recent"] = list(reversed(recent))

    robot = observation.get("robot") or {}
    robot_state: dict[str, Any] = {}
    currently_speaking = _optional_bool(robot.get("currentlySpeaking"), "robot.currentlySpeaking")
    if currently_speaking is not None:
        robot_state["currently_speaking"] = currently_speaking
    looking_at = robot.get("lookingAt")
    if looking_at is not None:
        if not isinstance(looking_at, str) or (looking_at != "none" and not _PERSON.fullmatch(looking_at)):
            raise ValueError("invalid robot.lookingAt")
        robot_state["looking_at"] = looking_at
    posture = _optional_choice(
        robot.get("posture"), ("idle", "attending", "nodding", "drooping"), "robot.posture"
    )
    if posture is not None:
        robot_state["posture"] = posture
    if _finite(robot.get("secondsSinceOwnLastTurn")):
        robot_state["seconds_since_own_last_turn"] = elapsed(robot["secondsSinceOwnLastTurn"])
    if robot_state:
        state["robot"] = robot_state
    return state

File: src/state.ts
export type Bearing = "far left" | "left" | "slightly left" | "center" | "slightly right" | "right" | "far right";
export type Distance = "very near" | "near" | "medium" | "far";
export type Elapsed = "just now" | "a few seconds" | "about 10 seconds" | "about half a minute" | "about a minute" | "over a minute";

export interface PersonObservation {
  id: string;
  bearingDeg?: number;
  faceHeightFraction?: number;
  faceYawDeg?: number;
  lookingAtRobot?: boolean;
  speaking?: boolean;
  secondsSinceLastSpoke?: number;
  neverSpoke?: boolean;
  moving?: "still" | "shifting" | "walking";
}
export interface RoomObservation {
  people?: PersonObservation[];
  sound?: { loudestBearingDeg?: number; levelDbfs?: number; voiceDetected?: boolean };
  transcriptRecent?: { who: string; text: string; endedSecondsAgo?: number }[];
  robot?: { currentlySpeaking?: boolean; lookingAt?: string; secondsSinceOwnLastTurn?: number; posture?: "idle" | "attending" | "nodding" | "drooping" };
}

function finite(value: number | undefined): value is number { return value !== undefined && Number.isFinite(value); }
function present<T>(value: T | null | undefined): value is T { return value !== undefined && value !== null; }
function record(value: unknown): boolean { return value !== null && typeof value === "object" && !Array.isArray(value); }
function optionalBoolean(value: unknown, field: string): boolean | undefined {
  if (!present(value)) return undefined;
  if (typeof value !== "boolean") throw new TypeError(`${field} must be boolean`);
  return value;
}
function optionalChoice<T extends string>(value: unknown, choices: readonly T[], field: string): T | undefined {
  if (!present(value)) return undefined;
  if (typeof value !== "string" || !choices.includes(value as T)) throw new TypeError(`invalid ${field}`);
  return value as T;
}
function optionalPersonTarget(value: unknown): string | undefined {
  if (!present(value)) return undefined;
  if (typeof value !== "string" || (value !== "none" && !/^p[1-9]$/.test(value))) throw new TypeError("invalid robot.lookingAt");
  return value;
}
function boundedTranscript(value: unknown): string {
  if (typeof value !== "string") throw new TypeError("transcript text must be string");
  let text = "";
  let units = 0;
  for (const scalar of value) {
    const codePoint = scalar.codePointAt(0)!;
    if (codePoint >= 0xd800 && codePoint <= 0xdfff) throw new TypeError("invalid transcript Unicode");
    if (units + scalar.length > 200) break;
    text += scalar;
    units += scalar.length;
  }
  return text;
}
export function bearing(degrees: number): Bearing {
  if (!Number.isFinite(degrees)) throw new RangeError("bearing must be finite");
  if (degrees < -60) return "far left";
  if (degrees < -25) return "left";
  if (degrees < -8) return "slightly left";
  if (degrees <= 8) return "center";
  if (degrees <= 25) return "slightly right";
  if (degrees <= 60) return "right";
  return "far right";
}
export function distance(fraction: number): Distance {
  if (!Number.isFinite(fraction) || fraction < 0 || fraction > 1) throw new RangeError("face height fraction must be in [0,1]");
  return fraction > 0.35 ? "very near" : fraction >= 0.20 ? "near" : fraction >= 0.10 ? "medium" : "far";
}
export function elapsed(seconds: number): Elapsed {
  if (!Number.isFinite(seconds) || seconds < 0) throw new RangeError("elapsed seconds must be non-negative");
  return seconds < 2 ? "just now" : seconds < 10 ? "a few seconds" : seconds < 20 ? "about 10 seconds" : seconds < 45 ? "about half a minute" : seconds < 90 ? "about a minute" : "over a minute";
}
export function soundLevel(dbfs: number): "silent" | "quiet" | "conversational" | "loud" {
  if (!Number.isFinite(dbfs) || dbfs > 0) throw new RangeError("dBFS must be finite and <= 0");
  return dbfs < -55 ? "silent" : dbfs < -35 ? "quiet" : dbfs < -15 ? "conversational" : "loud";
}
export function buildRoomState(input: RoomObservation) {
  if (!record(input) || (present(input.people) && !Array.isArray(input.people))
    || (present(input.sound) && !record(input.sound))
    || (present(input.robot) && !record(input.robot))
    || (present(input.transcriptRecent) && !Array.isArray(input.transcriptRecent))) throw new TypeError("invalid room observation shape");
  const personIds = new Set<string>();
  const people = (input.people ?? []).map((p) => {
    if (!record(p)) throw new TypeError("invalid person observation");
    const personId = p.id;
    if (typeof personId !== "string" || !/^p[1-9]$/.test(personId)) throw new TypeError("person IDs must be session-local p1..p9 identifiers");
    if (personIds.has(personId)) throw new TypeError("duplicate person ID");
    personIds.add(personId);
    const neverSpoke = optionalBoolean(p.neverSpoke, "person.neverSpoke");
    if (neverSpoke && present(p.secondsSinceLastSpoke)) throw new TypeError("neverSpoke conflicts with secondsSinceLastSpoke");
    const lookingAtRobot = optionalBoolean(p.lookingAtRobot, "person.lookingAtRobot");
    const speaking = optionalBoolean(p.speaking, "person.speaking");
    const moving = optionalChoice(p.moving, ["still", "shifting", "walking"], "person.moving");
    return {
      id: personId,
      ...(finite(p.bearingDeg) ? { bearing: bearing(p.bearingDeg) } : {}),
      ...(finite(p.faceHeightFraction) ? { distance: distance(p.faceHeightFraction) } : {}),
      ...(finite(p.faceYawDeg) ? { facing_robot: Math.abs(p.faceYawDeg) < 20 } : {}),
      ...(present(lookingAtRobot) ? { looking_at_robot: lookingAtRobot } : {}),
      ...(present(speaking) ? { speaking } : {}),
      ...(neverSpoke ? { seconds_since_last_spoke: "never" } : finite(p.secondsSinceLastSpoke) ? { seconds_since_last_spoke: elapsed(p.secondsSinceLastSpoke) } : {}),
      ...(present(moving) ? { moving } : {}),
    };
  });
  const voiceDetected = input.sound && optionalBoolean(input.sound.voiceDetected, "sound.voiceDetected");
  const sound = input.sound && {
    ...(finite(input.sound.loudestBearingDeg) ? { loudest_bearing: bearing(input.sound.loudestBearingDeg) } : {}),
    ...(finite(input.sound.levelDbfs) ? { level: soundLevel(input.sound.levelDbfs) } : {}),
    ...(present(voiceDetected) ? { voice_detected: voiceDetected } : {}),
  };
  const currentlySpeaking = input.robot && optionalBoolean(input.robot.currentlySpeaking, "robot.currentlySpeaking");
  const lookingAt = input.robot && optionalPersonTarget(input.robot.lookingAt);
  const posture = input.robot && optionalChoice(input.robot.posture, ["idle", "attending", "nodding", "drooping"], "robot.posture");
  const robot = input.robot && {
    ...(present(currentlySpeaking) ? { currently_speaking: currentlySpeaking } : {}),
    ...(present(lookingAt) ? { looking_at: lookingAt } : {}),
    ...(finite(input.robot.secondsSinceOwnLastTurn) ? { seconds_since_own_last_turn: elapsed(input.robot.secondsSinceOwnLastTurn) } : {}),
    ...(present(posture) ? { posture } : {}),
  };
  const latestBySpeaker = new Map<string, number>();
  if ((input.transcriptRecent ?? []).some((entry) => !record(entry))) throw new TypeError("invalid transcript observation");
  const transcript = [...(input.transcriptRecent ?? [])].reverse().filter(({ who }) => typeof who === "string" && (who === "unknown" || /^p[1-9]$/.test(who))).filter(({ who }) => {
    const count = latestBySpeaker.get(who) ?? 0;
    latestBySpeaker.set(who, count + 1);
    return count < 2;
  }).slice(0, 4).reverse().map(({ who, text, endedSecondsAgo }) => ({ who, text: boundedTranscript(text), ...(finite(endedSecondsAgo) ? { ended: elapsed(endedSecondsAgo) } : {}) }));
  return {
    schema: "room_state@1",
    people,
    ...(sound && Object.keys(sound).length ? { sound } : {}),
    ...(transcript.length ? { transcript_recent: transcript } : {}),
    ...(robot && Object.keys(robot).length ? { robot } : {}),
  };
}
